[Tokenizer Doc] Improve tokenizer summary (#8622)
huggingface/transformers: docs/source/tokenizer_summary.rst

> Summary of the tokenizers

Author: this is more in line with "Summary of the models" and "Summary of the tasks" in my opinion.

Summary of the tokenizers
-----------------------------------------------------------------------------------------------------------------------

On this page, we will have a closer look at tokenization. As we saw in :doc:`the preprocessing tutorial
<preprocessing>`, tokenizing a text is splitting it into words or subwords, which then are converted to ids through a
look-up table. Converting words or subwords to ids is straightforward, so in this summary, we will focus on splitting a
text into words or subwords (i.e. tokenizing a text). More specifically, we will look at the three main types of
tokenizers used in 🤗 Transformers: :ref:`Byte-Pair Encoding (BPE) <byte-pair-encoding>`, :ref:`WordPiece <wordpiece>`,
and :ref:`SentencePiece <sentencepiece>`, and show exemplary which tokenizer type is used by which model.

Note that on each model page, you can look at the documentation of the associated tokenizer to know which tokenizer

> type was used by the pre-trained model. For instance, if we look at :class:`~transformers.BertTokenizer`, we can see

Reviewer: ```suggestion
type was used by the pretrained model. For instance, if we look at :class:`~transformers.BertTokenizer`, we can see
```

that the model uses :ref:`WordPiece <wordpiece>`.

Introduction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Splitting a text into smaller chunks is a task that is harder than it looks, and there are multiple ways of doing so.
For instance, let's look at the sentence "Don't you love 🤗 Transformers? We sure do." A simple way of tokenizing this
text is to split it by spaces, which would give:

.. code-block::

    ["Don't", "you", "love", "🤗", "Transformers?", "We", "sure", "do."]

This is a sensible first step, but if we look at the tokens "Transformers?" and "do.", we notice that the punctuation
is attached to the words "Transformer" and "do", which is suboptimal. We should take the punctuation into account so
that a model does not have to learn a different representation of a word and every possible punctuation symbol that
could follow it, which would explode the number of representations the model has to learn. Taking punctuation into
account, tokenizing our exemplary text would give:

.. code-block::

>     ["Don", """, "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

Reviewer 2: If I'm not mistaken you got carried with a `:%s/'/"/g` :smile:

Author: that's very correct :D - corrected it

>     ["Don", """, "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

Reviewer: This change shouldn't be here.

>     ["Don", """, "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

Reviewer 2: ```suggestion
    ["Don", "'", "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]
```

Better. However, it is disadvantageous, how the tokenization dealt with the word "Don't". "Don't" stands for "do not",

> so it would be better tokenized as ``["Do", "n"t"]``. This is where things start getting complicated, and part of the

Reviewer 2: ```suggestion
so it would be better tokenized as ``["Do", "n't"]``. This is where things start getting complicated, and part of the
```

> so it would be better tokenized as ``["Do", "n"t"]``. This is where things start getting complicated, and part of the

Reviewer: ```suggestion
so it would be better tokenized as ``["Do", "n't"]``. This is where things start getting complicated, and part of the
```

reason each model has its own tokenizer type. Depending on the rules we apply for tokenizing a text, a different

> tokenized output is generated for the same text. A pre-trained model only performs properly if you feed it an input

Reviewer: ```suggestion
tokenized output is generated for the same text. A pretrained model only performs properly if you feed it an input
```

that was tokenized with the same rules that were used to tokenize its training data.

`spaCy <https://spacy.io/>`__ and `Moses <http://www.statmt.org/moses/?n=Development.GetStarted>`__ are two popular
rule-based tokenizers. Applying them on our example, *spaCy* and *Moses* would output something like:

.. code-block::

>     ["Do", "n"t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

Reviewer 2: ```suggestion
    ["Do", "n't", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]
```

> As can be seen Space -and punctuation tokenization, as well as rule-based tokenization, is used here. Space- and

Reviewer: ```suggestion
As can be seen space and punctuation tokenization, as well as rule-based tokenization, is used here. Space and
```

punctuation tokenization and rule-based tokenization are both examples of word tokenization, which is loosely defined
as splitting sentences into words. While it's the most intuitive way to split texts into smaller chunks, this

> tokenization method can lead to problems for massive text corpora. In this case, space- and punctuation tokenization

Reviewer: ```suggestion
tokenization method can lead to problems for massive text corpora. In this case, space and punctuation tokenization
```

> usually generates a very big vocabulary (the set of all unique words and tokens used). *E.g.*, :doc:`Transformer XL

Reviewer: ```suggestion
usually generates a very big vocabulary (the set of all unique words and tokens used). For instance, :doc:`Transformer XL
```

> <model_doc/transformerxl>` uses space- and punctuation-tokenization, resulting in a vocabulary size of 267,735!

Reviewer: ```suggestion
<model_doc/transformerxl>` uses space and punctuation tokenization, resulting in a vocabulary size of 267,735!
```
(or space-and-punctuation tokenization)

> Such a big vocabulary size forces the model to have an enormous embedding matrix as the input- and output layer, which

Reviewer: ```suggestion
Such a big vocabulary size forces the model to have an enormous embedding matrix as the input and output layer, which
```

> causes both an increased memory- and time complexity. In general, transformers models rarely have a vocabulary size

Reviewer: ```suggestion
causes both an increased memory and time complexity. In general, transformers models rarely have a vocabulary size
```

greater than 50,000, especially if they are pre-trained only on a single language.

> So if simple space- and punctuation-tokenization is unsatisfactory, why not simply tokenize on characters? While

Reviewer: ```suggestion
So if simple space and punctuation tokenization is unsatisfactory, why not simply tokenize on characters? While
```

> character tokenization is very simple and would greatly reduce memory- and time complexity it makes it much harder for

Reviewer: ```suggestion
character tokenization is very simple and would greatly reduce memory and time complexity it makes it much harder for
```

the model to learn meaningful input representations. *E.g.* learning a meaningful context-independent representation
for the letter "t" is much harder as learning a context-independent representation for the word "today". Therefore,
character tokenization is often accompanied by a loss of performance. So to get the best of both worlds, transformers
models use a hybrid between word-level and character-level tokenization called **subword** tokenization.

Subword tokenization
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Subword tokenization algorithms rely on the principle that frequently used words should not be split into smaller
subwords, but rare words should be decomposed into meaningful subwords. For instance "annoyingly" might be considered a

> rare word and could be decomposed into "annoying" and "ly". Both "annoying" and "ly" as stand-alone sub-words would

Reviewer: ```suggestion
rare word and could be decomposed into "annoying" and "ly". Both "annoying" and "ly" as stand-alone subwords would
```

appear more frequently while at the same time the meaning of "annoyingly" is kept by the composite meaning of
"annoying" and "ly". This is especially useful in agglutinative languages such as Turkish, where you can form (almost)
arbitrarily long complex words by stringing together subwords.

Subword tokenization allows the model to have a reasonable vocabulary size while being able to learn meaningful
context-independent representations. In addition, subword tokenization enables the model to process words it has never
seen before, by decomposing them into known subwords. For instance, the :class:`~transformers.BertTokenizer` tokenizes
"I have a new GPU!" as follows:

.. code-block::

    >>> from transformers import BertTokenizer
    >>> tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
    >>> tokenizer.tokenize("I have a new GPU!")
    ["i", "have", "a", "new", "gp", "##u", "!"]

Because we are considering the uncased model, the sentence was lowercased first. We can see that the words ``["i",

> "have", "a", "new"]`` are present in the tokenizer"s vocabulary, but the word "gpu" is not. Consequently, the tokenizer

Reviewer: ```suggestion
"have", "a", "new"]`` are present in the tokenizer"s vocabulary, but the word ``"gpu"`` is not. Consequently, the tokenizer
```
for consistency

> splits "gpu" into known subwords: ``["gp" and "##u"]``. "##" means that the rest of the token should be attached to the

Reviewer: ```suggestion
splits "gpu" into known subwords: ``["gp" and "##u"]``. ``"##"`` means that the rest of the token should be attached to the
```

Author: yes - applied to all standalone "<word>" occurances.

previous one, without space (for decoding or reversal of the tokenization).

As another example, :class:`~transformers.XLNetTokenizer` tokenizes our previously exemplary text as follows:

.. code-block::

    >>> from transformers import XLNetTokenizer
    >>> tokenizer = XLNetTokenizer.from_pretrained("xlnet-base-cased")
    >>> tokenizer.tokenize("Don't you love 🤗 Transformers? We sure do.")

>     ["▁Don", """, "t", "▁you", "▁love", "▁", "🤗", "▁", "Transform", "ers", "?", "▁We", "▁sure", "▁do", "."]

Reviewer 2: ```suggestion
    ["▁Don", "'", "t", "▁you", "▁love", "▁", "🤗", "▁", "Transform", "ers", "?", "▁We", "▁sure", "▁do", "."]
```

> We"ll get back to the meaning of those "▁" when we look at :ref:`SentencePiece <sentencepiece>`. As one can see, the

Reviewer: ```suggestion
We'll get back to the meaning of those "▁" when we look at :ref:`SentencePiece <sentencepiece>`. As one can see, the
```

Byte-Pair Encoding (BPE) was introduced in `Neural Machine Translation of Rare Words with Subword Units (Sennrich et
al., 2015) <https://arxiv.org/abs/1508.07909>`__. BPE relies on a pre-tokenizer that splits the training data into
words. Pretokenization can be as simple as space tokenization, e.g. :doc:`GPT-2 <model_doc/gpt2>`, :doc:`Roberta
<model_doc/roberta>`. More advanced pre-tokenization include rule-based tokenization, e.g. :doc:`XLM <model_doc/xlm>`,
:doc:`FlauBERT <model_doc/flaubert>` which uses Moses for most languages, or :doc:`GPT <model_doc/gpt>` which uses
Spacy and ftfy, to count the frequency of each word in the training corpus.

After pre-tokenization, a set of unique words has been created and the frequency of each word it occurred in the
training data has been determined. Next, BPE creates a base vocabulary consisting of all symbols that occur in the set
of unique words and learns merge rules to form a new symbol from two symbols of the base vocabulary. It does so until
the vocabulary has attained the desired vocabulary size. Note that the desired vocabulary size is a hyperparameter to
define before training the tokenizer.

> As an example, let"s assume that after pre-tokenization, the following set of words including their frequency has been

Reviewer: ```suggestion
As an example, let's assume that after pre-tokenization, the following set of words including their frequency has been
```

determined:

.. code-block::

    ("hug", 10), ("pug", 5), ("pun", 12), ("bun", 4), ("hugs", 5)

Consequently, the base vocabulary is ``["b", "g", "h", "n", "p", "s", "u"]``. Splitting all words into symbols of the
base vocabulary, we obtain:

.. code-block::

    ("h" "u" "g", 10), ("p" "u" "g", 5), ("p" "u" "n", 12), ("b" "u" "n", 4), ("h" "u" "g" "s", 5)

BPE then counts the frequency of each possible symbol pair and picks the symbol pair that occurs most frequently. In
the example above "h" followed by "u" is present `10 + 5 = 15` times (10 times in the 10 occurrences of "hug", 5 times
in the 5 occurrences of "hugs"). However, the most frequent symbol pair is "u" followed by "g", occurring `10 + 5 + 5 =
20` times in total. Thus, the first merge rule the tokenizer learns is to group all "u" symbols followed by a "g"
symbol together. Next, "ug" is added to the vocabulary. The set of words then becomes

.. code-block::

    ("h" "ug", 10), ("p" "ug", 5), ("p" "u" "n", 12), ("b" "u" "n", 4), ("h" "ug" "s", 5)

> BPE then identifies the next most common symbol pair. It"s "u" followed by "n", which occurs 16 times. "u", "n" is

Reviewer: ```suggestion
BPE then identifies the next most common symbol pair. It's "u" followed by "n", which occurs 16 times. "u", "n" is
```

is represented as

.. code-block::

    ("hug", 10), ("p" "ug", 5), ("p" "un", 12), ("b" "un", 4), ("hug" "s", 5)

Assuming, that the Byte-Pair Encoding training would stop at this point, the learned merge rules would then be applied
to new words (as long as those new words do not include symbols that were not in the base vocabulary). For instance,
the word "bug" would be tokenized to ``["b", "ug"]`` but "mug" would be tokenized as ``["<unk>", "ug"]`` since the
symbol "m" is not in the base vocabulary. In general, single letters such as "m" are not replaced by the "<unk>" symbol
because the training data usually includes at least one occurrence of each letter, but it is likely to happen for very
special characters like emojis.

As mentioned earlier, the vocabulary size, *i.e.* the base vocabulary size + the number of merges, is a hyperparameter
to choose. For instance :doc:`GPT <model_doc/gpt>` has a vocabulary size of 40,478 since they have 478 base characters
and chose to stop training after 40,000 merges.

Byte-level BPE
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> A base vocabulary that includes all possible base characters can be quite large if *e.g.* all unicode characters are

Reviewer: ```suggestion
A base vocabulary that includes all possible base characters can be quite large if for example all unicode characters are
```
clearer than e.g. here I think.

WordPiece is the subword tokenization algorithm used for :doc:`BERT <model_doc/bert>`, :doc:`DistilBERT
<model_doc/distilbert>`, and :doc:`Electra <model_doc/electra>`. The algorithm was outlined in `Japanese and Korean
Voice Seach (Schuster et al., 2012)
<https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/37842.pdf>`__ and is very similar to
BPE. WordPiece first initializes the vocabulary to include every character present in the training data and
progressively learn a given number of merge rules. In contrast to BPE, WordPiece does not choose the most frequent
symbol pair, but the one that maximizes the likelihood of the training data once added to the vocabulary.

So what does this mean exactly? Referring to the previous example, maximizing the likelihood of the training data is
equivalent to finding the symbol pair, whose probability divided by the probabilities of its first symbol followed by
its second symbol is the greatest among all symbol pairs. *E.g.* "u", followed by "g" would have only been merged if
the probability of "ug" divided by "u", "g" would have been greater than for any other symbol pair. Intuitively,

> WordPiece is slightly different to BPE in that it evaluates what it "loses" by merging two symbols to make ensure it's
> "worth it".

Reviewer: ```suggestion
WordPiece is slightly different to BPE in that it evaluates what it `loses` by merging two symbols to make ensure it's
`worth it`.
```
think italics is better here.
